### PR TITLE
Add canonicalization pass for doubly strided op 

### PR DESCRIPTION
**Pull request by jtuyls**
Adds a canonicalization pass for doubly strided operations. Reuses existing AIR DMA canonicalization logic.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir [hunk 1]
```diff
@@ -0,0 +1,250 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(func.func(iree-amdaie-canonicalize-doubly-strided-op,canonicalize))" %s | FileCheck %s
+
+// CHECK-LABEL: func.func @circular_dma_cpy_nd_source_target
```
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir:3]: I dont see in this test what canonicalization happened as the checks are same as the input IR?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+func.func @circular_dma_cpy_nd_partial_non_zero_offset(%arg0: !amdaie.logicalobjectfifo<memref<1x1x8x16xi32, 1>>, %arg1: !amdaie.logicalobjectfifo<memref<8x16xi32, 1>>) {
+  %0 = amdaie.circular_dma_cpy_nd(%arg0[0, 0, 0, 1] [1, 1, 8, 16] [128, 128, 16, 1], %arg1[0, 0, 0, 1] [1, 4, 2, 8] [64, 16, 8, 1]) : (!amdaie.logicalobjectfifo<memref<1x1x8x16xi32, 1>>, !amdaie.logicalobjectfifo<memref<8x16xi32, 1>>)
+  amdaie.logicalobjectfifo.consume(%0)
+  return
+}
+
+// -----
+
+// CHECK-LABEL: func.func @dma_cpy_nd_source_target
+// CHECK-SAME:  %[[ARG0:.+]]: !amdaie.logicalobjectfifo<memref<1x1x8x16xi32, 1>>
```
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir:87]: same comment as previous.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeDoublyStridedOp.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+void AMDAIECanonicalizeDoublyStridedOpPass::runOnOperation() {
+  Operation *parentOp = getOperation();
+  IRRewriter rewriter(parentOp->getContext());
+
+  // Fold linear dimensions within a DMA op.
+  parentOp->walk([&](AMDAIE::DoublyStridedOpInterface dmaOp) {
+    (void)foldDmaOpLinearDims(rewriter, dmaOp);
+  });
+
```
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeDoublyStridedOp.cpp:98]: IIRC, in the existing flow, the unit dim canonicalization was helping even in permuted cases where a dim was unit so could be dropped. Is something expected to be different in this flow that you anticipate not needing it or just haven't seen a use case yet?

**Review by nirvedhmeshram — CHANGES_REQUESTED** [review]
Looks great overall, had some minor comments and a question.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir:3]: Yeah, I am just verifying here that source/target is still correct after the transformation.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeDoublyStridedOp.cpp (hunk 3 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeDoublyStridedOp.cpp:98]: I haven't seen this use case yet. Let me have a look.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir (hunk 1 above)
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir:3]: It's good to add a comment saying the outputs are expected to be the same.

Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir:3]: Sure, will add it

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeDoublyStridedOp.cpp (hunk 3 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeDoublyStridedOp.cpp:128]: I added the unit dimension folding as well now: https://github.com/nod-ai/iree-amd-aie/pull/327/files#diff-9a02ada48f4391792b10c5eab979fe266d5c376ffdbc0bc52d5e34e3bf910212R130 + added specific testcases for this as well.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/canonicalize_doubly_strided_op.mlir:5]: Addded the comment

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+LogicalResult foldUnitDims(const SmallVector<OpFoldResult> &offsets,
+                           const SmallVector<OpFoldResult> &sizes,
+                           const SmallVector<OpFoldResult> &strides,
+                           SmallVector<OpFoldResult> &newOffsets,
+                           SmallVector<OpFoldResult> &newSizes,
+                           SmallVector<OpFoldResult> &newStrides) {
+  bool foldableUnitDimsFound = false;
+
+  for (int i = 0; i < offsets.size(); i++) {
+    // Dim can be folded if offset is zero and size is 1
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp:118]: Nit: use "0" instead of "zero".

**Review by yzhang93 — APPROVED** [review]
LGTM. Just several minor comments.

**Review by nirvedhmeshram — APPROVED** [review]
LGTM!

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+/// This describes accessing two times eight elements with a stride of eight in
+/// between. For example if the input is: [1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2,
+/// 2, 2, 2, 2] this will access [1, 1, 1, 1, 1, 1, 1, 1] and then [2, 2, 2, 2,
+/// 2, 2, 2, 2].
+///
+/// This is the same access pattern as accessing
+/// [1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2] in one go, which is
+/// represented by:
+///
+/// `offsets: [0], sizes: [32], strides: [1]`
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp]: Thanks for the catch. Updated it

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp (hunk 4 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp]: Thanks, done.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    return success(foldableLinearDimsFound);
+  }
+
+  newOffsets.push_back(offsets[0]);
+  newStrides.push_back(strides[0]);
+  newSizes.push_back(sizes[0]);
+
+  for (int i = 1; i < offsets.size(); i++) {
+    // Conditions for folding a dim.
+    // 1. size(i) x stide(i) == stride(i-1), with this we can have new size(i-1)
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp]: Thanks, done.

**Pull request merged**